fix(EngineService): persist controller state changes during Engine.init

andrepimenta opened the pull request:
## **Description**

Controllers can modify state during `Engine.init()` (e.g., `SnapController` granting permissions to `PermissionController` when handling preinstalled snaps). These state changes were not being persisted because persistence subscriptions are set up **after** controller construction completes, missing any events emitted during initialization.

CHANGELOG entry: null

### The Problem

1. `EngineService.start()` calls `Engine.init()`
2. During `Engine.init()`, controllers are constructed via `initModularizedControllers()`
3. `SnapController` constructor calls `#handlePreinstalledSnaps()` → `#updatePermissions()` → `PermissionController:grantPermissions`
4. `PermissionController` updates its state (e.g., adds snap permissions)
5. `PermissionController` emits `stateChange` event, but no subscriptions exist yet
6. After `Engine.init()` returns, persistence subscriptions are set up - **too late to capture the changes**

### The Solution

This fix passes the initial state (loaded from persistence before init) to `setupEnginePersistence()` and compares controller state after init using **1-level deep referential equality**.

Controllers preserve referential equality for unchanged properties (via Immer). If a property reference at the first level differs from the initial state, the state changed during init and needs to be persisted.

### Key Changes

- Add `initialState` parameter to `initializeControllers()` and `setupEnginePersistence()`
- Compare `filteredState` properties against `initialState` at 1-level deep
- Handle new installs/controllers where `initialState` is undefined
- Only persist if there's actual state to save (non-empty `filteredState`)
- Reuse the same debounced `persistController` function for consistency

## **Changelog**

CHANGELOG entry: Fixed an issue where controller state changes during engine initialization were not being persisted

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: Persist controller state changes during Engine.init()

  Scenario: Preinstalled snap permissions are persisted on fresh install
    Given the app is freshly installed with no persisted state
    
    When the app starts and Engine.init() completes
    Then PermissionController state should contain preinstalled snap permissions
    And the PermissionController state should be persisted to storage
    
    When the app is killed and restarted
    Then the preinstalled snap permissions should still be present

  Scenario: Controller state changes during init are detected and persisted
    Given the app has persisted state from a previous session
    
    When the app starts and a controller's state changes during Engine.init()
    Then the changed controller state should be persisted
    And a log message should indicate the controller state was queued for persist
```

## **Screenshots/Recordings**

### **Before**

N/A - This is a persistence/data fix, not a UI change.

### **After**

N/A - This is a persistence/data fix, not a UI change.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Detects and persists controller state changes occurring during Engine.init by comparing post-init filtered state to initial state, and adds comprehensive tests.
> 
> - **EngineService**
>   - Pass `initialState` into `initializeControllers` and `setupEnginePersistence` to detect changes made during `Engine.init()`.
>   - In `setupEnginePersistence`, compute filtered state via `getPersistentState` and compare 1-level deep against `initialState`; if changed and non-empty, persist immediately using debounced `persistController` and log.
>   - Maintain subscriptions for future state changes; add error handling/logging and skip controllers without persistent state.
>   - Update `start()` and `initializeVaultFromBackup()` to pass loaded state when initializing controllers.
> - **Tests**
>   - Add/expand tests covering: init-time persistence detection, unchanged state, new installs, property removals, empty filtered state, missing `controllerMessenger`, and graceful error handling.
>   - Ensure cleanup/restoration of mocked properties and avoid unhandled rejections in persistence tests.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b0cf0da45093344a804a86e716fee93dac4bfac5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

andrepimenta committed:
fix(EngineService): persist controller state changes during Engine.init()

Controllers can modify state during Engine.init() (e.g., SnapController
granting permissions to PermissionController). These changes were not
being persisted because persistence subscriptions are set up after
controller construction completes.

This fix passes the initial state to setupEnginePersistence and compares
controller state using 1-level deep referential equality. Controllers
preserve referential equality for unchanged properties (via Immer), so
if a property reference differs, the state changed and needs persistence.

Changes:
- Add initialState parameter to initializeControllers and setupEnginePersistence
- Compare filteredState properties against initialState at 1-level deep
- Handle new installs/controllers where initialState is undefined
- Only persist if there's actual state to save (non-empty filteredState)
- Reuse the same debounced persistController function for consistency
- Skip CronjobController (handled separately by CronjobControllerStorageManager)
- Update initializeVaultFromBackup to pass state for consistency

andrepimenta committed:
test(EngineService): add tests for state change detection during init

Add tests to verify the new behavior that detects and persists controller
state changes that occur during Engine.init().

New tests:
- persists state that changed during Engine.init()
- does not persist state that has not changed during Engine.init()
- persists state for new install when no initial state exists
- does not persist empty state for controllers without state to save

Fixed existing tests:
- logs persistence errors without crashing: reorder mock setup to avoid
  unhandled promise rejections from init-time state change detection
- handles missing controllerMessenger: restore original descriptor after
  test to prevent state leakage to subsequent tests

MarioAslau reviewed (dismissed):
Looks good to me!

andrepimenta committed:
fix(EngineService): guard against undefined currentState in init persistence

Add guard check and try-catch around init-time state change detection
to prevent app crashes when controller state is undefined.

- Add truthy check for currentState before calling getPersistentState
- Wrap entire state change detection block in try-catch for safety
- Ensure subscription setup always runs regardless of detection errors
- Log errors but don't crash - init-time persistence is best-effort

Gudahtt commented on app/core/EngineService/EngineService.ts:
```diff
@@ -215,8 +236,61 @@ export class EngineService {
+                      filteredState[key] !==
+                      initialControllerState[
+                        key as keyof typeof initialControllerState
+                      ],
```
SonarCloud seems to be correct here, the type assertion is not needed

```suggestion
                    (key) => filteredState[key] !== initialControllerState[key],
```

Gudahtt reviewed (dismissed):
LGTM!

andrepimenta committed:
Minor improvements

andrepimenta committed:
fix(EngineService): detect property removal in state change comparison

The previous logic only iterated over keys in filteredState, missing cases
where properties were removed during init. For example, if initial state
was { a: 1, b: 2 } and current state is { a: 1 }, the change wasn't detected.

Now compares key counts to detect added/removed properties in addition to
checking for value changes.

andrepimenta commented on app/core/EngineService/EngineService.ts:
```diff
@@ -215,8 +231,67 @@ export class EngineService {
+
+              if (hasChanged) {
+                // Reuse the same debounced function for consistency
+                persistController(filteredState, controllerName);
```
We don't want to await for persist to finish before we set the listeners

andrepimenta committed:
test(EngineService): add test for property removal detection during init

Adds test case to verify that when a property is removed from controller
state during Engine.init() (e.g., { a: 1, b: 2 } -> { a: 1 }), the change
is correctly detected by comparing key counts and persistence is triggered.
